libs/execution_context/ebpf_native.c

```diff
@@ -468,6 +647,97 @@ _ebpf_native_release_reference_internal(void* base_object, ebpf_file_id_t file_i
     _ebpf_native_release_reference(base_object);
 }
 
+static ebpf_result_t
+_ebpf_native_validate_native_module(_In_opt_ const metadata_table_t* table, _In_ const GUID* client_module_id)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    metadata_table_t local_table = {0};
+    bpf2c_version_t version;
+    program_entry_t* programs = NULL;
+    size_t program_count = 0;
+    map_entry_t* maps = NULL;
+    size_t map_count = 0;
+    map_initial_values_t* map_initial_values = NULL;
+    size_t map_initial_values_count = 0;
+    global_variable_section_info_t* global_variables = NULL;
+    size_t global_variable_count = 0;
+
+    if (!table || !table->programs || !table->maps) {
+        EBPF_LOG_MESSAGE_GUID(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_NATIVE,
+            "The metadata table size is invalid",
+            client_module_id);
+        return EBPF_INVALID_ARGUMENT;
+    }
+
+    if (table->size < EBPF_OFFSET_OF(metadata_table_t, version)) {
+        EBPF_LOG_MESSAGE_GUID(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_NATIVE,
+            "The metadata table size is invalid",
+            client_module_id);
+        return EBPF_INVALID_ARGUMENT;
+    }
+
+    // Copy the metadata table.
+    memcpy(&local_table, table, min(table->size, sizeof(metadata_table_t)));
+
+    // Check if the client module is compatible with the runtime.
+    if (local_table.size < EBPF_OFFSET_OF(metadata_table_t, version)) {
+        result = EBPF_INVALID_ARGUMENT;
+        EBPF_LOG_MESSAGE_GUID(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_NATIVE,
+            "The metadata table size is wrong for client module. The version of bpf2c used to generate this module "
+            "may be too old.",
+            client_module_id);
+        goto Done;
+    }
+
+    local_table.version(&version);
+    if (_ebpf_compare_versions(&version, &_ebpf_minimum_version) < 0) {
+        result = EBPF_INVALID_ARGUMENT;
+        EBPF_LOG_MESSAGE_GUID(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_NATIVE,
+            "Native module version is older than _ebpf_minimum_version. Rejecting the attach request.",
+            client_module_id);
+        goto Done;
+    }
+
+    // Validate the programs.
+    local_table.programs(&programs, &program_count);
+    if (!_ebpf_validate_native_program_entry_array(programs, program_count)) {
+        result = EBPF_INVALID_ARGUMENT;
```

Comment:
Do we want to trace error statements for each of these checks? Or would there be sufficient diagnostic information without this?

Reply:
Added.